Feature/#3 category

HTH016 opened the pull request:
Setting
- build.gradle, setting.gradle git ignore 에서 제외

JWT 
- ROLE claim 추가

Security
- swagger 설정 추가
- url 권한 설정

카테고리 기능 추가 
- admin 권한만 접근 가능
- 카테고리 추가, 목록 조회, 삭제, 복구
- 관리자는 active 상태인 카테고리와 soft deleted 상태인 카테고리를 구분하여 볼 수 있습니다.  

Api Response Convention 변경
- SuccessResponse, ErrorResponse 구분
- Controller 에 반영

DTO 이름 변경

Test code 작성

HTH016 committed:
카테고리 기능 구현
카테고리 추가, 삭제, 복구, 조회
JWT Claim Role 추가

HTH016 committed:
카테고리 기능 구현
카테고리 추가, 삭제, 복구, 조회
JWT Claim Role 추가

HTH016 committed:
카테고리 기능 구현
카테고리 추가, 삭제, 복구, 조회
JWT Claim Role 추가

HTH016 committed:
#01 수정
primary key를 Long 으로 변경

HTH016 committed:
#01 수정
primary key를 Long 으로 변경

HTH016 committed:
Merge branch 'feature/#1-signup-login' into feature/#3-category

HTH016 committed:
#03 (수정)
API Response Wrapping
Test Code
DTO 파일명 변경

HTH016 committed:
#03 (수정)
API Response Wrapping
Test Code
DTO 파일명 변경

mason1155 commented on src/test/java/flab/Linkedlog/LinkedLogApplicationTests.java:
```diff
@@ -1,13 +1,153 @@
+    private CategoryService categoryService;
+
+    @Test
+        // 카테고리 추가
```
주석 대신 DisplayName를 써주세요

mason1155 commented on src/test/java/flab/Linkedlog/categoryTest/DeleteCategoryTests.java:
```diff
@@ -0,0 +1,226 @@
+                .andExpect(jsonPath("$.success").value(true));
+
+        Category restoredCategory = categoryRepository.findById(categoryId).orElseThrow();
+        assert restoredCategory.getDeletedAt() == null;
```
junit5, assertj 라이브러리를 공부해보세요 
위에서 쓰신 assertThat 등을 써주세요

mason1155 commented on src/main/java/flab/Linkedlog/service/CategoryService.java:
```diff
@@ -0,0 +1,113 @@
+
+    // 카테고리 삭제(숨김) (관리자)
+    public void deleteCategory(CategoryDeleteRequest categoryDeleteRequest) {
+        Category category = categoryRepository.findById(categoryDeleteRequest.getId())
```
deletedAt이 null 인걸 조회해주세요
그리고 jpa의 @where 어노테이션 찾아보시죠

mason1155 commented on src/main/java/flab/Linkedlog/controller/CategoryController.java:
```diff
@@ -0,0 +1,157 @@
+                    .response(errorResponse)
+                    .build();
+
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(apiResponse);
```
RestControllerAdvice 찾아보고 적용해주세요

mason1155 commented on src/main/java/flab/Linkedlog/controller/CategoryController.java:
```diff
@@ -0,0 +1,157 @@
+    private final CategoryService categoryService;
+
+    @PostMapping(value = "/admin/addcategory")
+    public ResponseEntity<ApiResponse<?>> addCategory(@RequestBody @Validated CategoryCreateRequest categoryCreateRequest) {
```
ResponseEntity 생략가능하니 응답 타입에서 제외해주세요

mason1155 commented on src/main/java/flab/Linkedlog/controller/response/ApiResponse.java:
```diff
@@ -0,0 +1,16 @@
+@Builder
+public class ApiResponse<T> {
+    private boolean success;
+    private T response;
```
하나의 응답 파일에서 성공 응답 바디를 모두 포함하는 방식과 성공 응답, 실패 응답을 분리하는 방법 중에 하나 선택해서 수정해주세요
지금처럼 응답이 깊이가 길어지는 것은 지양해주세요

mason1155 commented on src/main/java/flab/Linkedlog/config/JwtAuthenticationFilter.java:
```diff
@@ -24,11 +26,15 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
+
             try {
                 String username = jwtUtil.validateToken(token);
+                String role = (String) jwtUtil.getClaimsFromToken(token).get("roles");
```
token에 role이 없을 수 도 있지 않을까요? 없으면 형변환시 에러발생할 걸로 보여요

mason1155 commented on src/main/java/flab/Linkedlog/config/SecurityConfiguration.java:
```diff
@@ -32,14 +38,38 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
     }
 
+    @Bean
+    public OpenAPI customOpenAPI() {
```
OpenAPI는 다른 설정 클래스로 빼시죠

mason1155 commented on src/main/java/flab/Linkedlog/controller/CategoryController.java:
```diff
@@ -0,0 +1,157 @@
+
+    @GetMapping(value = "/admin/categoryList")
+    public ResponseEntity<ApiResponse<?>> getCategories(
+            @RequestParam(name = "status", defaultValue = "active") String status) {
```
String보단 Enum으로 받으시죠

mason1155 commented on src/main/java/flab/Linkedlog/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,55 @@
+    List<Category> findAllDeleted();
+
+    // 숨김, 삭제 해제 (복구)
+    default void restoreCategory(Category category) {
```
굳이 여기에 default 메서드로 만들 필요는 없어보이요

HTH016 committed:
#03 (수정)
Test Code AssertThat 사용
ApiResponse depth 변경, Error response를 GlobalExceptionHandler 로 관리
SwaggerConfig 분리
CategoryState Enum 추가

HTH016 committed:
#03 (수정)
Test Code AssertThat 사용
ApiResponse depth 변경, Error response를 GlobalExceptionHandler 로 관리
SwaggerConfig 분리
CategoryState Enum 추가

HTH016 committed:
#03 (수정)
사용하지 않는 Test 파일 제외

mason1155 commented on src/main/java/flab/Linkedlog/controller/response/ApiResponse.java:
```diff
@@ -0,0 +1,36 @@
+    private boolean success;
+    private T response;
+    private String error;
+    private String message;
```
응답 필드를 4개에서 2개로 줄이고 null인 값은 응답에서 제외하는 어노테이션이 있으니 찾아서 적용해보시죠

HTH016 committed:
#03 (수정)
category delete 테스트 추가
ApiResponse 필드 변경
카테고리 삭제, 복구 중복확인 코드 추가

mason1155 reviewed (approved):
고생하셨어요

mason1155 commented on src/test/java/flab/Linkedlog/categoryTest/AddCategoryTests.java:
```diff
@@ -0,0 +1,120 @@
+                        .header("Authorization", "Bearer " + adminToken)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(request)))
+                .andReturn();
```
andExpect으로 상태 코드, 응답 바디 검증하셔도 됩니다.
